### PR TITLE
Delete pushers after calling on_logged_out module hook on device delete

### PR DESCRIPTION
**Pull request by MatMaul**
The assumption that `on_logged_out` module hooks will be called before deleting the pushers has been "broken" by #15280.

This behavior was not really specified, hence my quoted `broken`, so this PR also add it to the doc of the module API.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: docs/modules/password_auth_provider_callbacks.md [hunk 1]
```diff
@@ -103,6 +103,9 @@ Called during a logout request for a user. It is passed the qualified user ID, t
 deactivated device (if any: access tokens are occasionally created without an associated
 device ID), and the (now deactivated) access token.
 
+Deleting the related pushers is done after calling `on_logged_out`, so you can rely on them
+to still be present.
```
Comment (clokep) [inline docs/modules/password_auth_provider_callbacks.md:107]: Can we add a test so we don't break it again?

Comment (MatMaul) [reply docs/modules/password_auth_provider_callbacks.md:107]: Good point, will do.

File: tests/module_api/test_api.py [hunk 2]
```diff
@@ -773,6 +774,49 @@ def test_create_room(self) -> None:
         # Check room alias.
         self.assertIsNone(room_alias)
 
+    def test_on_logged_out(self) -> None:
+        """Test that on_logged_out module hook is properly call when logging out
+        a device, and that related pushers are still available at this time.
+        """
```
Comment (clokep) [inline tests/module_api/test_api.py:780]:
```suggestion
        """Test that the related pushers are still available when the on_logged_out module hook is called.
        """
```

File: tests/module_api/test_api.py [hunk 3]
```diff
@@ -773,6 +774,49 @@ def test_create_room(self) -> None:
...
+        self.assertEqual(self.nb_of_pushers_in_callback, 1)
+
+        self.assertEqual(len(self.hs.get_pusherpool().pushers[user_id].values()), 0)
+
+    async def _on_logged_out_mock(
+        self, user_id: str, device_id: Optional[str], access_token: str
+    ) -> None:
+        self.nb_of_pushers_in_callback = len(
+            self.hs.get_pusherpool().pushers[user_id].values()
+        )
```
Comment (clokep) [inline tests/module_api/test_api.py:818]: I think this is clearer if we don't hang state on the test case object:

```suggestion
        # Setup a callback to count the number of pushers.
        number_of_pushers_in_callback: Optional[int] = None

	    async def on_logged_out_mock(
	        self, user_id: str, device_id: Optional[str], access_token: str
	    ) -> None:
	        nonlocal number_of_pushers_in_callback
	        number_of_pushers_in_callback = len(
	            self.hs.get_pusherpool().pushers[user_id].values()
	        )

        self.module_api.register_password_auth_provider_callbacks(
            on_logged_out=on_logged_out_mock
        )

        # Delete the device.
        device_handler = self.hs.get_device_handler()
        assert isinstance(device_handler, DeviceHandler)
        self.get_success(device_handler.delete_devices(user_id, [device_id]))

        # Check that the callback was called and the pushers still existed.
        self.assertEqual(self.nb_of_pushers_in_callback, 1)

        # Ensure the pushers were deleted after the callback.
        self.assertEqual(len(self.hs.get_pusherpool().pushers[user_id].values()), 0)
```

**Review by clokep — COMMENTED** [review]
Just some nits!

**Review by clokep — APPROVED** [review]
Thanks!

**Pull request merged**